File: tests/generic/gen_cor_062/Input/event_listener.py
from web3 import Web3
import logging
import requests, asyncio
import argparse, json, sys
from eth_account.messages import encode_defunct

logging.basicConfig(format='%(asctime)s %(message)s', stream=sys.stdout, level=logging.INFO)


def generate_viewing_key(web3, url, address, private_key):
    logging.info('Generating viewing key for %s' % private_key)

    headers = {'Accept': 'application/json', 'Content-Type': 'application/json'}
    data = {"address": address}
    response = requests.post('%s/generateviewingkey/' % url, data=json.dumps(data), headers=headers)

    signed_msg = web3.eth.account.sign_message(encode_defunct(text='vk' + response.text), private_key=private_key)
    data = {"signature": signed_msg.signature.hex(), "address": address}
    requests.post('%s/submitviewingkey/' % url, data=json.dumps(data), headers=headers)


async def log_loop(contract, address, poll_interval):
    last_balance = 0
    while True:
        balance = contract.functions.balanceOf(address).call()
        if balance > last_balance:
            last_balance = balance
            logging.info('New balance = %s' % balance)
        await asyncio.sleep(poll_interval)


def main(contract, address):
    logging.info('Starting to run the event loop')
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(
            asyncio.gather(
                log_loop(contract, address, 1)))
    finally:
        loop.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(prog='event_listener')
    parser.add_argument("url", help="Connection URL")
    parser.add_argument("address", help="Address of the contract")
    parser.add_argument("abi", help="Abi of the contract")
    parser.add_argument("pk", help="Private key of account to poll")
    args = parser.parse_args()

    logging.info('URL: %s' % args.url)
    logging.info('ADR: %s' % args.address)
    logging.info('ABI: %s' % args.abi)
    logging.info('PK: %s' % args.pk)

    web3 = Web3(Web3.HTTPProvider(args.url))
    account = web3.eth.account.privateKeyToAccount(args.pk)
    if args.pk: generate_viewing_key(web3, args.url, account.address, args.pk)
    with open(args.abi) as f:
        contract = web3.eth.contract(address=args.address, abi=json.load(f))
    main(contract, account.address)

File: tests/generic/gen_cor_076/Input/event_listener.js
const Web3 = require('web3');
const commander = require('commander');
const fs = require('fs');

require('console-stamp')(console, 'HH:MM:ss');

function task(web3, topic) {
  console.log('Starting task ...')
  web3.eth.subscribe('logs', {topics: ['0x12345...']},
    function(error, result) {
      if (error) {
        console.log('Error returned is ', error)
      } else {
        console.log('Stored value =', result.returnValues['value']);
      }
    }
  )
}

function generate_viewing_key(web3, contract, url, account, private_key) {
  console.log('Generating viewing key for', private_key)
  console.log(url+'/generateviewingkey/')

  fetch(url+'/generateviewingkey/', {
    method: 'POST',
    headers: {'Accept': 'application/json', 'Content-Type': 'application/json'},
    body: JSON.stringify({address: account.address})
  })
  .then(response => response.text())
  .then((response) => {
         sign_viewing_key(web3, contract, url, account, private_key, response)
   })
}

function sign_viewing_key(web3, contract, url, account, private_key, response) {
  console.log('Signing viewing key for', private_key)
  console.log('Result was', response)
  signed_msg = web3.eth.accounts.sign('vk' + response, '0x' + private_key)

  fetch(url+'/submitviewingkey/', {
    method: 'POST',
    headers: {'Accept': 'application/json', 'Content-Type': 'application/json'},
    body: JSON.stringify( {signature: signed_msg.signature, address: account.address})
  })
  .then(response => response.text())
  .then((response) => {
    task(contract, account.address)
   })
}

commander
  .version('1.0.0', '-v, --version')
  .usage('[OPTIONS]...')
  .option('-u, --url_http <url>', 'Http connection URL')
  .option('-w, --url_ws <url>', 'Web socket connection URL')
  .option('-a, --address <value>', 'Address of the contract')
  .option('-b, --abi <value>', 'Abi of the contract')
  .option('-p, --pk <value>', 'Private key of account to poll')
  .option('-o, --obscuro', 'True if running against obscuro', false)
  .parse(process.argv);

const options = commander.opts();
console.log('HTTP URL:', `${options.url_http}`);
console.log('WS URL:', `${options.url_ws}`);
console.log('ADR:', `${options.address}`);
console.log('ABI:', `${options.abi}`);
console.log('PK:', `${options.pk}`);
console.log('OB:', `${options.obscuro}`);

const json = fs.readFileSync(`${options.abi}`);
const abi = JSON.parse(json);

const web3 = new Web3(`${options.url_ws}`);
const contract = new web3.eth.Contract(abi, `${options.address}`)
account = web3.eth.accounts.privateKeyToAccount(`${options.pk}`)

if (options.obscuro == true) {
  generate_viewing_key(web3, contract, options.url_http, account, options.pk)
}
else {
  task(contract, account.address)
}










